### PR TITLE
FAST - Fixes 498 - 02-networking-nva stage

### PR DESCRIPTION
**Pull request by LucaPrete**
Folks, I drop here this PR before the weekend. Let's review it together carefully, as it touches important pieces.

FYI:

1. fixed GCP ASN defaults: it was conflicting with the peer ASN. Moved to 64512 as in networking-vpn
2. added a table summarizing the subnets in the readme
3. for subnets conflicts: changes spoke subnet cidr, PSA ranges and l7ILB ranges (ip subnetting [here](https://www.davidc.net/sites/default/subnets/subnets.html?network=10.128.0.0&mask=16&division=63.727539a62ba94531)):

- the spoke subnets now consume a /20 instead of a /19
- for each spoke, the other /20 has been further subnetted, to obtain (also) 3x /24 (more below)
- the last /24 of each range is used for the corresponding regional spoke l7ilb subnet
- the second-last /24 of each range is used for PSA sql server regional spoke subnet
- the third-last /24 of each range is used for PSA mysql server regional spoke subnet

### CODE REVIEW

**Comment by LucaPrete** [comment]
Closes #498

**Comment by sruffilli** [comment]
Hey Luca, variable `custom_adv` doesn't list subnets, but aggregate ranges as they should be routed/advertised, meaning you should carve your ip space from that range.

So, using `gcp_dev_ew1="10.128.128.0/20"` as an example:

- `dev-default-ew1` subnet should be a /24
- `var.l7ilb_subnets.dev` should use the third to last available /24 for ew1 (i.e. `{ ip_cidr_range = "10.128.141.0/24", region = "europe-west1" }`)
- `var.psa_ranges.dev` should use the second to last and last available /24 for ew1 
```
    dev = {
      cloudsql-mysql-ew1     = "10.128.142.0/24"
      cloudsql-sqlserver-ew1 = "10.128.143.0/24"
#[...]
    }
```

LMK if it makes sense.

File: fast/stages/02-networking-nva/data/subnets/dev/dev-default-ew1.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # skip boilerplate check
 
 region: europe-west1
-ip_cidr_range: 10.128.128.0/19
+ip_cidr_range: 10.128.128.0/20
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/data/subnets/dev/dev-default-ew1.yaml:4]: Per my comment, individual subnets are by default /24 carved from the IP space assigned to a specific spoke/region.

File: fast/stages/02-networking-nva/data/subnets/dev/dev-default-ew4.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # skip boilerplate check
 
 region: europe-west4
-ip_cidr_range: 10.128.160.0/19
+ip_cidr_range: 10.128.160.0/20
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/data/subnets/dev/dev-default-ew4.yaml:4]: Per my comment, individual subnets are by default /24 carved from the IP space assigned to a specific spoke/region.

File: fast/stages/02-networking-nva/data/subnets/prod/prod-default-ew1.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # skip boilerplate check
 
 region: europe-west1
-ip_cidr_range: 10.128.192.0/19
+ip_cidr_range: 10.128.192.0/20
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/data/subnets/prod/prod-default-ew1.yaml:4]: Per my comment, individual subnets are by default /24 carved from the IP space assigned to a specific spoke/region.

File: fast/stages/02-networking-nva/data/subnets/prod/prod-default-ew4.yaml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # skip boilerplate check
 
 region: europe-west4
-ip_cidr_range: 10.128.224.0/19
+ip_cidr_range: 10.128.224.0/20
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/data/subnets/prod/prod-default-ew4.yaml:4]: Per my comment, individual subnets are by default /24 carved from the IP space assigned to a specific spoke/region.

File: fast/stages/02-networking-nva/variables.tf [hunk 5]
```diff
@@ -26,14 +26,14 @@ variable "custom_adv" {
   default = {
     cloud_dns                 = "35.199.192.0/19"
     gcp_all                   = "10.128.0.0/16"
-    gcp_dev_ew1               = "10.128.128.0/19"
-    gcp_dev_ew4               = "10.128.160.0/19"
+    gcp_dev_ew1               = "10.128.128.0/20"
+    gcp_dev_ew4               = "10.128.160.0/20"
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/variables.tf:30]: I'd either use /20 or /19 everywhere

File: fast/stages/02-networking-nva/variables.tf [hunk 6]
```diff
@@ -69,13 +69,13 @@ variable "l7ilb_subnets" {
...
   })))
   default = {
-    prod = [
-      { ip_cidr_range = "10.128.92.0/24", region = "europe-west1" },
-      { ip_cidr_range = "10.128.93.0/24", region = "europe-west4" }
-    ]
     dev = [
-      { ip_cidr_range = "10.128.60.0/24", region = "europe-west1" },
-      { ip_cidr_range = "10.128.61.0/24", region = "europe-west4" }
+      { ip_cidr_range = "10.128.159.0/24", region = "europe-west1" },
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/variables.tf:73]: Per my comment, use the third to last /24 available on the carved IP estate.
Also applies to all following ranges.

File: fast/stages/02-networking-nva/variables.tf [hunk 7]
```diff
@@ -121,13 +121,17 @@ variable "psa_ranges" {
...
   type        = map(map(string))
   default = {
-    prod = {
-      cloudsql-mysql     = "10.128.94.0/24"
-      cloudsql-sqlserver = "10.128.95.0/24"
-    }
     dev = {
-      cloudsql-mysql     = "10.128.62.0/24"
-      cloudsql-sqlserver = "10.128.63.0/24"
+      cloudsql-mysql-ew1     = "10.128.157.0/24"
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/variables.tf:125]: Per my comment, use the secondo to last and last /24 available on the carved IP estate.

**Review by sruffilli — CHANGES_REQUESTED** [review]

**Comment by LucaPrete** [comment]
Simo, makes sense. Good catch!

Why at this point we don’t simply advertise the whole /19 aggregate (which
is indeed dedicated to each “regional spoke”?

Il giorno ven 4 feb 2022 alle 18:18 Simone Ruffilli <
***@***.***> ha scritto:

**Comment by sruffilli** [comment]
> Simo, makes sense. Good catch! Why at this point we don’t simply advertise the whole /19 aggregate (which is indeed dedicated to each “regional spoke”?

Yep, that's the idea.

**Comment by LucaPrete** [comment]
I made the changes requested.

We need to be aware that this way we'll be advertising to on-prem the l7ilb subnets as well, which won't hurt, but it's probably not what we (/users?) would like to do.

Should we use a completely different subnet for different the l7ilb subnets, so we won't announce them?

**Comment by sruffilli** [comment]
I don't think you've pushed what you think you've pushed :) 
Also, please enjoy your weekend!

File: fast/stages/02-networking-nva/README.md [hunk 8]
```diff
@@ -81,36 +81,53 @@ This is an options summary:
...
 ### IP ranges, subnetting, routing
 
-These are the subnets allocated by default in this setup:
+Minimizing the number of routes (and subnets) in the cloud environment is important, as it simplifies management and it avoids hitting [Cloud Router](https://cloud.google.com/network-connectivity/docs/router/quotas) and [VPC](https://cloud.google.com/vpc/docs/quota) quotas and limits. For this reason, we recommend to carefully plan the IP space used in your cloud environment. This allows the use of larger IP CIDR blocks in routes, whenever possible.
+
+This stage uses a dedicated /16 block (10.128.0.0/16), which should be sized to the own needs. The subnets created in each VPC derive from this range.
+
+The /16 block is evenly split in eight, smaller /19 blocks, assigned to different areas of the GCP network: *landing untrusted europe-west1*, *landing untrusted europe-west4*, *landing trusted europe-west1*, *landing untrusted europe-west4*, *development europe-west1*, *development europe-west4*, *production europe-west1*, *production europe-west4*.
+
+The first /24 range is carved from each /19 and effectively allocated for the subnet.
```
Comment (sruffilli) [inline fast/stages/02-networking-nva/README.md:90]: `The first /24 range is carved from each /19 and effectively allocated for the subnet.`
->
`The first /24 on every area is allocated for a default subnet, which can be removed or modified as needed`

**Review by sruffilli — APPROVED** [review]

**Comment by sruffilli** [comment]
LGTM modulo the required rebase and the tiny update in the README

**Pull request merged**